### DISCUSSION TITLE
Best method/code for sleeping Heltec WSL(V3) device?

### DISCUSSION THREAD

**Discussion by michapr** [jgromes/RadioLib#1079] [Q&A] [answered]
Hi,

I have tested the actual Radiolib with the actual board description files from Heltec (v.3.0.0) and have used the sleeping sample from here: https://github.com/radiolib-org/radiolib-persistence

Have modified the `gotoSleep()`:
```
void gotoSleep(uint32_t seconds) {
  esp_sleep_enable_timer_wakeup(seconds * 1000UL * 1000UL); // function uses uS
  Serial.println(F("Sleeping\n"));
  Serial.flush();
  Serial.end();
  SPI.end();
  radio.sleep();

  esp_deep_sleep_start();

  // if this appears in the serial debug, we didn't go to sleep!
  // so take defensive action so we don't continually uplink
  Serial.println(F("\n\n### Sleep failed, delay of 5 minutes & then restart ###\n"));
  delay(5UL * 60UL * 1000UL);
  ESP.restart();
}
```

Anyway I do not get a good sleeping current - all what I have tried is about 800μA.
Have tried to set additional the pins to analog input, or to other values, but no difference.

It must be possible - because if I use the original Heltec sample code I get about **30μA** (!!) sleeping current (same timer wakeup).
Not sure what this can be - maybe the radio.sleep() is there different (pins of SX1262 have othe sleep state) or something else...
Board is powered by a Lithium cell.

My question - somebody have an idea, or made a good sleeping device with WSLV3 and actual RadioLib?

BTW: with new board description this init of SPI is needed (after updating all (lib and board files) nothing was working more... :( ...) 
 -> `  SPI.begin(9, 11, 10, 8);`

**Comment by StevenCellist** [comment]
Configure all pins that are connected to the radio as `ANALOG` using `pinMode()`

**Reply by StevenCellist** [reply]
Alternatively, I find this snippet of code in one of my tests:
```cpp
     pinMode(8, OUTPUT);
     digitalWrite(8, HIGH);
     gpio_hold_en((gpio_num_t)8);
```
... which effectively holds the NSS/CS pin high during sleep. Currently not setup for power measurement, so can't tell if that is actually the solution here.

**Reply by michapr** [reply]
Thanks for this, but it doesn't help - same results about (`ANALOG` have also tried before....). 
gpio_hold_en I have also tried before, because it was helping using the RFM95W with ESP32C3. Here I got about 90μA sleeping current (the voltage regulator take about 40μA). (Also a bit more as expected - but wirering there must be optimized (noise on wires))

So with WSLV3 must be something else possible....

**Reply by StevenCellist** [reply, marked as answer]
How about putting the radio to sleep _before_ ending SPI?

**Reply by michapr** [reply]
YES, of course, my mistake. Have added the `SPI.end();` after some tries while testing at the wrong place.

Attentive eyes, - I had a look at the code 1000 times - and haven't seen it...

Get now 14-16μA !

THANK YOU!

**Reply by StevenCellist** [reply]
Now if you can get the power consumption below 100uA when using a GPIO wakeup (like through ext0/ext1), please let me know..
